File: src/org/jetbrains/plugins/scala/lang/parser/ScalaParserDefinition.java
package org.jetbrains.plugins.scala.lang.parser;

import com.intellij.lang.ParserDefinition;
import com.intellij.lang.PsiParser;
import com.intellij.lang.ASTNode;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.project.Project;
import com.intellij.psi.tree.IFileElementType;
import com.intellij.psi.tree.TokenSet;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.FileViewProvider;
import org.jetbrains.annotations.NotNull;

/**
 * Author: Ilya Sergey
 * Date: 25.09.2006
 * Time: 14:23:22
 */
public class ScalaParserDefinition implements ParserDefinition {
    @NotNull                 
    public Lexer createLexer(Project project) {
        throw new UnsupportedOperationException("createLexer not implemented in org.jetbrains.plugins.scala.lang.parser.ScalaParserDefinition");
    }

    public PsiParser createParser(Project project) {
        throw new UnsupportedOperationException("createParser not implemented in org.jetbrains.plugins.scala.lang.parser.ScalaParserDefinition");
    }

    public IFileElementType getFileNodeType() {
        throw new UnsupportedOperationException("getFileNodeType not implemented in org.jetbrains.plugins.scala.lang.parser.ScalaParserDefinition");
    }

    @NotNull
    public TokenSet getWhitespaceTokens() {
        throw new UnsupportedOperationException("getWhitespaceTokens not implemented in org.jetbrains.plugins.scala.lang.parser.ScalaParserDefinition");
    }

    @NotNull
    public TokenSet getCommentTokens() {
        throw new UnsupportedOperationException("getCommentTokens not implemented in org.jetbrains.plugins.scala.lang.parser.ScalaParserDefinition");
    }

    @NotNull
    public PsiElement createElement(ASTNode astNode) {
        throw new UnsupportedOperationException("createElement not implemented in org.jetbrains.plugins.scala.lang.parser.ScalaParserDefinition");
    }

    public PsiFile createFile(FileViewProvider fileViewProvider) {
        throw new UnsupportedOperationException("createFile not implemented in org.jetbrains.plugins.scala.lang.parser.ScalaParserDefinition");
    }

    public SpaceRequirements spaceExistanceTypeBetweenTokens(ASTNode astNode, ASTNode astNode1) {
        throw new UnsupportedOperationException("spaceExistanceTypeBetweenTokens not implemented in org.jetbrains.plugins.scala.lang.parser.ScalaParserDefinition");
    }
}

File: src/org/jetbrains/plugins/scala/lang/parser/ScalaParserDefinition.scala
package org.jetbrains.plugins.scala
package lang
package parser

import com.intellij.lang.{ASTNode, ParserDefinition}
import com.intellij.openapi.project.Project
import com.intellij.psi.stubs.PsiFileStub
import com.intellij.psi.tree.{IStubFileElementType, TokenSet}
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.psi.{FileViewProvider, PsiElement, PsiFile}
import org.jetbrains.plugins.scala.lang.lexer.{ScalaLexer, ScalaTokenTypes}
import org.jetbrains.plugins.scala.lang.psi.api.toplevel.imports.ScImportStmt
import org.jetbrains.plugins.scala.lang.psi.impl.ScalaFileImpl
import org.jetbrains.plugins.scala.settings._

/**
  * @author ilyas
  */
class ScalaParserDefinition extends ScalaParserDefinitionWrapper {

  protected val psiCreator: PsiCreator = ScalaPsiCreator

  def createLexer(project: Project): ScalaLexer = {
    val settings = ScalaProjectSettings.getInstance(project)
    new ScalaLexer(settings.isTreatDocCommentAsBlockComment)
  }

  def createParser(project: Project): ScalaParser =
    new ScalaParser

  def getFileNodeType: IStubFileElementType[_ <: PsiFileStub[_ <: PsiFile]] =
    ScalaElementTypes.FILE

  def getCommentTokens: TokenSet = ScalaTokenTypes.COMMENTS_TOKEN_SET

  def getStringLiteralElements: TokenSet = ScalaTokenTypes.STRING_LITERAL_TOKEN_SET

  def getWhitespaceTokens: TokenSet = ScalaTokenTypes.WHITES_SPACES_TOKEN_SET

  def createElement(astNode: ASTNode): PsiElement =
    psiCreator.createElement(astNode)

  def createFile(fileViewProvider: FileViewProvider): PsiFile = {
    ScalaFileFactory.EP_NAME.getExtensions
      .view
      .flatMap(_.createFile(fileViewProvider))
      .headOption
      .getOrElse(new ScalaFileImpl(fileViewProvider))
  }

  override def spaceExistanceTypeBetweenTokens(leftNode: ASTNode, rightNode: ASTNode): ParserDefinition.SpaceRequirements = {
    val importStatement = PsiTreeUtil.getParentOfType(leftNode.getPsi, classOf[ScImportStmt])
    val isNeighbour = Option(importStatement).exists {
      _.getTextRange.getEndOffset == rightNode.getTextRange.getStartOffset
    }

    import com.intellij.lang.ParserDefinition.SpaceRequirements._
    rightNode.getElementType match {
      case ScalaTokenTypes.tWHITE_SPACE_IN_LINE if rightNode.getText.contains('\n') => MAY
      case _ if isNeighbour => MUST_LINE_BREAK
      case ScalaTokenTypes.kIMPORT => MUST_LINE_BREAK
      case _ => super.spaceExistanceTypeBetweenTokens(leftNode, rightNode)
    }
  }
}
